llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp

```diff
@@ -1605,11 +1703,14 @@ bool SIWholeQuadMode::runOnMachineFunction(MachineFunction &MF) {
 
   // Shader is simple does not need any state changes or any complex lowering
   if (!(GlobalFlags & (StateWQM | StateStrict)) && LowerToCopyInstrs.empty() &&
-      LowerToMovInstrs.empty() && KillInstrs.empty()) {
+      LowerToMovInstrs.empty() && KillInstrs.empty() &&
+      InitExecInstrs.empty()) {
```

Comment:
Could you just add `lowerInitExecInstrs()` to this code path and avoid the empty check.
This way there is no need to run the rest of the pass for just an SI_INIT_EXEC?

Reply:
Done. But really, as a future cleanup, I'd prefer to remove this fast path. Instead we should make sure that the "slow" path below is actually fast when all the `Instrs` vectors are empty, and returns an accurate "Changed" flag.